File: flutter_inappwebview_forge_web/lib/web/web_view_lifecycle_coordinator.dart
/// Internal lifecycle state machine for an iframe-backed WebView.
///
/// This type is intentionally not exported from the package library. It keeps
/// ownership and callback admission rules in one place without changing the
/// public Dart API or the platform channel contract.
enum WebViewLifecycleState {
  creating,
  preparing,
  attached,
  ready,
  detachedRetained,
  reattached,
  rendererLost,
  disposing,
  disposed,
}

class WebViewLifecycleCoordinator {
  WebViewLifecycleState state = WebViewLifecycleState.creating;

  int _operationId = 0;
  int _pendingAsyncOperations = 0;
  int _callbackCompletionCount = 0;
  final Set<int> _activeOperationIds = <int>{};
  List<String>? _debugTrace;

  bool get isDisposed => state == WebViewLifecycleState.disposed;

  bool get acceptsCallbacks =>
      state != WebViewLifecycleState.disposing && !isDisposed;

  bool beginPreparing() {
    if (!acceptsCallbacks || state == WebViewLifecycleState.ready) {
      return false;
    }
    if (state != WebViewLifecycleState.creating &&
        state != WebViewLifecycleState.reattached &&
        state != WebViewLifecycleState.detachedRetained) {
      return false;
    }
    state = WebViewLifecycleState.preparing;
    _record('preparing');
    return true;
  }

  void markAttached() {
    if (!acceptsCallbacks) {
      return;
    }
    if (state == WebViewLifecycleState.detachedRetained ||
        state == WebViewLifecycleState.rendererLost) {
      state = WebViewLifecycleState.reattached;
    } else if (state == WebViewLifecycleState.creating ||
        state == WebViewLifecycleState.preparing) {
      state = WebViewLifecycleState.attached;
    }
    _record('attached');
  }

  void markReady() {
    if (acceptsCallbacks &&
        (state == WebViewLifecycleState.attached ||
            state == WebViewLifecycleState.reattached ||
            state == WebViewLifecycleState.preparing)) {
      state = WebViewLifecycleState.ready;
      _record('ready');
    }
  }

  void markDetachedRetained() {
    if (acceptsCallbacks) {
      state = WebViewLifecycleState.detachedRetained;
      _record('detachedRetained');
    }
  }

  void markReattached() {
    if (acceptsCallbacks) {
      state = WebViewLifecycleState.reattached;
      _record('reattached');
    }
  }

  bool markRendererLost() {
    if (!acceptsCallbacks) {
      return false;
    }
    state = WebViewLifecycleState.rendererLost;
    _record('rendererLost');
    return true;
  }

  int get operationId => _operationId;

  int get pendingAsyncOperations => _pendingAsyncOperations;

  int get callbackCompletionCount => _callbackCompletionCount;

  int? beginAsyncOperation() {
    if (!acceptsCallbacks) {
      return null;
    }
    _pendingAsyncOperations++;
    _operationId++;
    _activeOperationIds.add(_operationId);
    return _operationId;
  }

  bool completeAsyncOperation(int operationId) {
    if (!_activeOperationIds.remove(operationId)) {
      return false;
    }
    _pendingAsyncOperations--;
    _callbackCompletionCount++;
    return true;
  }

  bool beginDisposal() {
    if (state == WebViewLifecycleState.disposing || isDisposed) {
      return false;
    }
    state = WebViewLifecycleState.disposing;
    _operationId++;
    _record('disposing');
    return true;
  }

  void finishDisposal() {
    _activeOperationIds.clear();
    _pendingAsyncOperations = 0;
    state = WebViewLifecycleState.disposed;
    _record('disposed');
  }

  void _record(String event) {
    assert(() {
      (_debugTrace ??= <String>[]).add(event);
      return true;
    }());
  }
}

File: flutter_inappwebview_forge_android/android/src/main/kotlin/com/emirkanacar/flutter_inappwebview_forge_android/types/WebViewLifecycleCoordinator.kt
package com.emirkanacar.flutter_inappwebview_forge_android.types

/**
 * Internal lifecycle state for one native WebView.
 *
 * This is deliberately not exposed through the Flutter API. It gives native
 * callbacks and asynchronous operations one shared disposal boundary while
 * preserving the existing channel contract.
 */
internal enum class WebViewLifecycleState {
    CREATING,
    PREPARING,
    ATTACHED,
    READY,
    DETACHED_RETAINED,
    REATTACHED,
    RENDERER_LOST,
    DISPOSING,
    DISPOSED
}

internal class WebViewLifecycleCoordinator {
    private var debugTrace: MutableList<String>? = null
    private val activeOperationIds: MutableSet<Long> = HashSet()

    @Volatile
    var state: WebViewLifecycleState = WebViewLifecycleState.CREATING
        private set

    @Volatile
    var operationId: Long = 0
        private set

    @Volatile
    var pendingAsyncOperations: Int = 0
        private set

    @Volatile
    var callbackCompletionCount: Long = 0
        private set

    @Synchronized
    fun beginPreparing(): Boolean {
        if (state != WebViewLifecycleState.CREATING) return false
        state = WebViewLifecycleState.PREPARING
        record("preparing")
        return true
    }

    @Synchronized
    fun markAttached() {
        when (state) {
            WebViewLifecycleState.PREPARING -> {
                state = WebViewLifecycleState.ATTACHED
                record("attached")
            }
            WebViewLifecycleState.DETACHED_RETAINED,
            WebViewLifecycleState.RENDERER_LOST -> {
                state = WebViewLifecycleState.REATTACHED
                record("reattached")
            }
            else -> Unit
        }
    }

    @Synchronized
    fun markReady() {
        if (state == WebViewLifecycleState.ATTACHED ||
            state == WebViewLifecycleState.PREPARING ||
            state == WebViewLifecycleState.REATTACHED
        ) {
            state = WebViewLifecycleState.READY
            record("ready")
        }
    }

    @Synchronized
    fun markDetachedRetained() {
        if (state == WebViewLifecycleState.READY ||
            state == WebViewLifecycleState.ATTACHED ||
            state == WebViewLifecycleState.REATTACHED
        ) {
            state = WebViewLifecycleState.DETACHED_RETAINED
            record("detachedRetained")
        }
    }

    @Synchronized
    fun markRendererLost(): Boolean {
        if (!acceptsCallbacksLocked()) return false
        state = WebViewLifecycleState.RENDERER_LOST
        record("rendererLost")
        return true
    }

    @Synchronized
    fun beginAsyncOperation(): Long? {
        if (!acceptsCallbacksLocked()) return null
        pendingAsyncOperations += 1
        operationId += 1
        activeOperationIds.add(operationId)
        return operationId
    }

    @Synchronized
    fun completeAsyncOperation(operationId: Long): Boolean {
        if (!activeOperationIds.remove(operationId)) return false
        pendingAsyncOperations -= 1
        callbackCompletionCount += 1
        return true
    }

    @Synchronized
    fun beginDisposal(): Boolean {
        if (state == WebViewLifecycleState.DISPOSING ||
            state == WebViewLifecycleState.DISPOSED
        ) {
            return false
        }
        state = WebViewLifecycleState.DISPOSING
        operationId += 1
        record("disposing")
        return true
    }

    @Synchronized
    fun finishDisposal() {
        activeOperationIds.clear()
        pendingAsyncOperations = 0
        state = WebViewLifecycleState.DISPOSED
        record("disposed")
    }

    @Synchronized
    fun acceptsCallbacks(): Boolean = acceptsCallbacksLocked()

    private fun acceptsCallbacksLocked(): Boolean =
        state != WebViewLifecycleState.DISPOSING &&
            state != WebViewLifecycleState.DISPOSED

    @Synchronized
    internal fun enableDebugTrace() {
        debugTrace = mutableListOf()
    }

    @Synchronized
    internal fun lifecycleTrace(): List<String> = debugTrace?.toList() ?: emptyList()

    private fun record(event: String) {
        debugTrace?.add(event)
    }
}
